### PR TITLE
Commit the script for TestLink Migration

### PR DESCRIPTION
**Pull request by TGavan**
This is the pull request for the python script which help in migrating the test cases from TestLink to TestRail

### CODE REVIEW

File: testlink-migration/.gitignore [hunk 1]
```diff
@@ -0,0 +1,5 @@
+.DS_Store
+**/node_modules
+**/target
+use_cases/graphql_api_usage/java
```
Comment (bitcoder) [inline testlink-migration/.gitignore]: I don't think this makes sense here

File: testlink-migration/.gitignore [hunk 2]
```diff
@@ -0,0 +1,5 @@
+.DS_Store
+**/node_modules
```
Comment (bitcoder) [inline testlink-migration/.gitignore]: I don't think this makes sense here

File: testlink-migration/.gitignore [hunk 3]
```diff
@@ -0,0 +1,5 @@
...
+**/target
```
Comment (bitcoder) [inline testlink-migration/.gitignore]: I don't think this makes sense here

File: testlink-migration/LICENSE
Comment (bitcoder) [inline testlink-migration/LICENSE:1]: this file need to be removed.. besides, it has mentions to another product

File: testlink-migration/TestLink2TestRail.py [hunk 4]
```diff
@@ -0,0 +1,160 @@
+import sys, getopt
+import re
+import xml.etree.ElementTree as ET
+import pandas as pd
+
+column = ["Issue ID","Issue Key","Test Type","Test Summary", "Test Priority", "Action","Data","Result","Test Set","Precondition","Issue Type", "Precondition Type"]
```
Comment (bitcoder) [inline testlink-migration/TestLink2TestRail.py]: I don't think the following fields are relevant for TestRail right? And they won't have any data in TestRail case.. or am I seeing it wrong?

"Test Set","Precondition","Issue Type", "Precondition Type"

File: testlink-migration/TestLink2TestRail.py [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+                "Issue Type": 'testset',
+                "Precondition Type": ''})
+
+def handleTestSuites(root, issueID, outputfile):
+    # When exporting from a Test Suite Testlink adds one level
+    testsuites = root.findall('testsuite')
+    if root.tag == 'testsuite' and len(testsuites) <= 1:
+        # Parse Testcases
+        name = root.attrib['name']
+        appendTestset(issueID,testSet=name)
```
Comment (bitcoder) [inline testlink-migration/TestLink2TestRail.py]: there is no TestSet concept and we won't process it right? If so, please remove this

File: testlink-migration/TestLink2TestRail.py [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+        issueID=handleTestCases(root, issueID+1, issueID, outputfile=outputfile)
+    else:
+        for testsuite in testsuites:
+            # Parse Testcases
+            name = testsuite.attrib['name']
+            appendTestset(issueID,testSet=name)
```
Comment (bitcoder) [inline testlink-migration/TestLink2TestRail.py]: there is no TestSet concept and we won't process it right? If so, please remove this

File: testlink-migration/TestLink2TestRail.py [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+            issueID=handleTestCases(testsuite, issueID+1, issueID, outputfile=outputfile)
+
+def handleTestCases(root, issueID, testsetid, outputfile):
```
Comment (bitcoder) [inline testlink-migration/TestLink2TestRail.py]: if won't process the testsetid, then remove that parameter

File: testlink-migration/TestLink2TestRail.py [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+def parseTestlink2TestRailData(inputfile, outputfile):
+    # Parsing XML file
+    xmlParse = ET.parse(inputfile)
+    root = xmlParse.getroot()
+    issueID = 1
+
+    if root.tag != 'testcase' and root.tag != 'testcases':
+        handleTestSuites(root=root, issueID=issueID, outputfile=outputfile)
+    else:
+        handleTestCases(root=root, issueID=issueID, testsetid='', outputfile=outputfile)
```
Comment (bitcoder) [inline testlink-migration/TestLink2TestRail.py]: i we won't process the testsetid, then pls remove that argument

File: testlink-migration/TestLink2TestRail.py [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+            action = step.find('actions').text
+            expectedResult = step.find('expectedresults').text
+            if first_step:
+                appendRows(issueID=issueID,testType=test_Type,testSummary=summary,testPriority=test_priority,action=action,result=expectedResult,precondition=preconditionID, testSet=testsetid)
+                first_step = False
+            else:
+                appendRows(issueID=issueID,testType=test_Type, action=action,result=expectedResult)
+        
+        if not hasSteps:
+            appendRows(issueID=issueID,testType=test_Type,testSummary=summary,testPriority=test_priority,precondition=preconditionID, testSet=testsetid)
```
Comment (bitcoder) [inline testlink-migration/TestLink2TestRail.py]: if we won't process the testsetid or the preconditionId , then pls remove those arguments

File: testlink-migration/TestLink2TestRail.py [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+    if root.tag != 'testcase':
+        root = root.findall('testcase')
+
+    for testcase in root:
+        # Parse Testcases
+        name = testcase.attrib['name']
+        summary = testcase.find('summary').text
+        precondition = testcase.find('preconditions').text
+        if precondition:
+            appendPrecondition(issueID=issueID, precondition=precondition)
```
Comment (bitcoder) [inline testlink-migration/TestLink2TestRail.py]: are we going to do anything with the precondition?

File: testlink-migration/TestLink2TestRail.py [hunk 11]
```diff
@@ -0,0 +1,160 @@
...
+                "Test Priority": '',
+                "Action": '',
+                "Data": '',
+                "Result": '',
+                "Test Set": '',
+                "Precondition": '',
+                "Issue Type": 'precondition',
+                "Precondition Type": 'Manual'})
+
+def appendTestset(issueID='', testSet=''):
```
Comment (bitcoder) [inline testlink-migration/TestLink2TestRail.py]: there is no TestSet concept and we won't process it right? If so, please remove this

File: testlink-migration/TestLink2TestRail.py [hunk 12]
```diff
@@ -0,0 +1,160 @@
...
+                "Test Priority": testPriority,
+                "Action": cleanTags(action),
+                "Data": '',
+                "Result": cleanTags(result),
+                "Test Set": testSet if testSet else '',
+                "Precondition": precondition if precondition else '',
+                "Issue Type": issueType,
+                "Precondition Type": ''})
+
+def appendPrecondition(issueID='', precondition=''):
```
Comment (bitcoder) [inline testlink-migration/TestLink2TestRail.py]: if we won't process Precondition , please remove this

File: testlink-migration/.gitignore (hunk 1 above)
Comment (TGavan) [reply testlink-migration/.gitignore]: @bitcoder its now  removed

File: testlink-migration/.gitignore (hunk 2 above)
Comment (TGavan) [reply testlink-migration/.gitignore]: @bitcoder its now removed

File: testlink-migration/.gitignore (hunk 3 above)
Comment (TGavan) [reply testlink-migration/.gitignore]: @bitcoder its now removed

File: testlink-migration/LICENSE
Comment (TGavan) [reply testlink-migration/LICENSE:1]: @bitcoder file has been removed

File: testlink-migration/TestLink2TestRail.py (hunk 4 above)
Comment (TGavan) [reply testlink-migration/TestLink2TestRail.py]: @bitcoder its now removed

File: testlink-migration/TestLink2TestRail.py (hunk 8 above)
Comment (TGavan) [reply testlink-migration/TestLink2TestRail.py]: @bitcoder its now removed

File: testlink-migration/TestLink2TestRail.py (hunk 9 above)
Comment (TGavan) [reply testlink-migration/TestLink2TestRail.py]: @bitcoder its removed now

File: testlink-migration/TestLink2TestRail.py (hunk 10 above)
Comment (TGavan) [reply testlink-migration/TestLink2TestRail.py]: @bitcoder its now removed

File: testlink-migration/TestLink2TestRail.py (hunk 11 above)
Comment (TGavan) [reply testlink-migration/TestLink2TestRail.py]: @bitcoder its removed now

File: testlink-migration/TestLink2TestRail.py (hunk 12 above)
Comment (TGavan) [reply testlink-migration/TestLink2TestRail.py]: @bitcoder its now removed

**Review by bitcoder — APPROVED** [review]

**Pull request merged**